pkg/controller/kubelet-config/kubelet_config_nodes.go

```diff
@@ -0,0 +1,336 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/kubelet/config/v1beta1"
+
+	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	ctrlcommon "github.com/openshift/machine-config-operator/pkg/controller/common"
+	"github.com/openshift/machine-config-operator/pkg/version"
+)
+
+func (ctrl *Controller) nodeConfigWorker() {
+	for ctrl.processNextNodeConfigWorkItem() {
+	}
+}
+
+func (ctrl *Controller) processNextNodeConfigWorkItem() bool {
+	key, quit := ctrl.nodeConfigQueue.Get()
+	if quit {
+		return false
+	}
+	defer ctrl.nodeConfigQueue.Done(key)
+
+	err := ctrl.syncNodeConfigHandler(key.(string))
+	ctrl.handleNodeConfigErr(err, key)
+	return true
+}
+
+func (ctrl *Controller) handleNodeConfigErr(err error, key interface{}) {
+	if err == nil {
+		ctrl.nodeConfigQueue.Forget(key)
+		return
+	}
+
+	if ctrl.nodeConfigQueue.NumRequeues(key) < maxRetries {
+		glog.V(2).Infof("Error syncing node configuration %v: %v", key, err)
+		ctrl.nodeConfigQueue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	glog.V(2).Infof("Dropping node config %q out of the queue: %v", key, err)
+	ctrl.nodeConfigQueue.Forget(key)
+	ctrl.nodeConfigQueue.AddAfter(key, 1*time.Minute)
+}
+
+// syncNodeConfigHandler syncs whenever there is a change on the nodes.config.openshift.io resource
+// nodes.config.openshift.io object holds the cluster-wide information about the
+// node specific features such as cgroup modes, workerlatencyprofiles, etc.
+func (ctrl *Controller) syncNodeConfigHandler(key string) error {
+	startTime := time.Now()
+	glog.V(4).Infof("Started syncing nodeConfig handler %q (%v)", key, startTime)
+	defer func() {
+		glog.V(4).Infof("Finished syncing nodeConfig handler %q (%v)", key, time.Since(startTime))
+	}()
+
+	// Fetch the Node
+	nodeConfig, err := ctrl.nodeConfigLister.Get(ctrlcommon.ClusterNodeInstanceName)
+	if errors.IsNotFound(err) {
+		glog.V(2).Infof("Node configuration %v is missing, using default", key)
+		nodeConfig = createNewDefaultNodeconfig()
+	} else if err != nil {
+		glog.V(2).Infof("%v", err)
+		err := fmt.Errorf("could not fetch Node: %v", err)
+		return err
+	}
+	// Find all MachineConfigPools
+	mcpPools, err := ctrl.mcpLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, pool := range mcpPools {
+		role := pool.Name
+		// the configuration change will be applied only on the worker nodes.
+		if role == mcfgv1.MachineConfigPoolMaster {
+			continue
+		}
+		// Get KubeletConfig
+		key, err := getManagedNodeKey(pool, ctrl.client)
+		if err != nil {
+			return err
+		}
+		kc, err := ctrl.client.MachineconfigurationV1().KubeletConfigs().Get(context.TODO(), key, metav1.GetOptions{})
+		isNotFound := errors.IsNotFound(err)
+		if err != nil && !isNotFound {
+			return err
+		}
+		if isNotFound {
+			ignConfig := ctrlcommon.NewIgnConfig()
+			kc, err = ctrlcommon.KubeletConfigFromIgnConfig(role, key, ignConfig)
```

Comment:
I'm a bit confused as to this general flow. Am I understanding correctly that you create a kubeletconfig for the nodes resource under a specific key?

Since we are the kubeletconfigcontroller, shouldn't we be creating a new managed MachineConfig (generated via kubeletconfig if necessary in the controller) for nodes resources?

Reply:
Another thing to consider is that each pool can only have 1 kubeletconfig. Meaning that you will either:
1. merge into user created ones (don't do this)
2. overwrite user created ones or have user created ones overwrite you, neither of which is an option

Reply:
> I'm a bit confused as to this general flow. Am I understanding correctly that you create a kubeletconfig for the nodes resource under a specific key?
> 
> Since we are the kubeletconfigcontroller, shouldn't we be creating a new managed MachineConfig (generated via kubeletconfig if necessary in the controller) for nodes resources?

@yuqi-zhang , I understand the concern and totally agree with your suggestion. My initial set of code changes were to create/update an MC whenever there is a change in the node config object. The only reason that I modified the code to create/update KC is to actually rely on it for the status updates.

I would update the code to create/update MC and would now rely on the corresponding pool status for updating the node config object's status. Should that be okay?

Reply:
Hi Sai,

So to confirm, does the changes from the `node.config.openshift.io` modify the base kubelet files? If so, then I think the logic would have to be such that:

1. if there is the base system templates, create a 98-xxx MachineConfig (like we have today for featuregates) that merges in the base configs, maybe around https://github.com/openshift/machine-config-operator/blob/master/pkg/controller/kubelet-config/kubelet_config_controller.go#L594 ?
2. if there is a featuregate, merge the featuregate changes and the nodes changes together with the base into 98-xx
3. if there is a user kubeletconfig, merge all of the above into 99-xxx
4. if there is a new user kubeletconfig, merge the new one, the featuregates, the node changes, and the base templates into 99-xxx-2

Basically, we need to make sure that features/featuregates/nodes config/user kubeletconfigs all get rendered together based on what combination of those exist. There must only be 1 MachineConfig for 

Conversely, if changes to node.config.openshift.io do NOT have to be merged into the base configs, this is much easier. You can e.g. create a MC that you manage that creates dropin files with unique names instead. This won't conflict.

Happy to discuss this synchronously if that would be helpful. Feel free to reach out to me on slack anytime, thanks!